File: crates/yokan/demo/points.py
# /// script
# requires-python = ">=3.14"
# ///
"""struct ↔ frozen dataclass. `frozen=True` is the admission ticket:
an immutable value cannot expose Python's reference aliasing, so it
means the same thing as a native COW value by construction. Updates
are `dataclasses.replace` — a new value, both tiers.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from dataclasses import dataclass, replace  # noqa: E402

from yokan import button, column, row, run, State, text  # noqa: E402


@dataclass(frozen=True)
class Point:
    x: int
    y: int = 0


sel: State[Point] = State(Point(3, 4))
dist: State[int] = State(0)


def move_right():
    sel.set(replace(sel(), x=sel().x + 5))


def swap():
    sel.set(Point(sel().y, sel().x))


def measure():
    p = sel()
    dist.set(p.x * p.x + p.y * p.y)


def view():
    with column(spacing=8, padding=12):
        text(f"p=({sel().x}, {sel().y}) d2={dist()}")
        with row(spacing=6):
            button("right", on_click=move_right)
            button("swap", on_click=swap)
            button("measure", on_click=measure)


if __name__ == "__main__":
    run(view, title="points")

File: rakugan/demo/points.pl
# A small class of values, carried on the app's own state. A class with
# no `view` is a value: `:param` says what `new` is given, `:reader`
# what can be read back, and the compiled run holds it as a value
# rather than as something two names can share.
use Rakugan;

class Point {
    use Rakugan;
    field $x :param :reader = 0;
    field $y :param :reader = 0;
}

class Points {
    use Rakugan;
    field $sel  = Point->new(x => 3, y => 4);
    field $dist = 0;

    method view {
        return column(
            text("p=(@{[ $sel->x ]}, @{[ $sel->y ]}) d2=$dist"),
            row(
                button("right",   on_click => sub { $sel = Point->new(x => $sel->x + 5, y => $sel->y) }),
                button("swap",    on_click => sub { $sel = Point->new(x => $sel->y, y => $sel->x) }),
                button("measure", on_click => sub { $dist = $sel->x * $sel->x + $sel->y * $sel->y }),
                spacing => 6,
            ),
            spacing => 8,
            padding => 12,
        );
    }
}

run(Points->new, title => "points");
